### PR TITLE
Reserved connection streaming

### PR DESCRIPTION
**Pull request by harshit-gangal**
## Description

This PR adds support for modifying system settings, allows advisory locking in streaming(OLAP) mode.

## Checklist
- [X] Tests were added or are not required
- [X] Documentation was added or is not required

### CODE REVIEW

File: go/vt/vtgate/scatter_conn.go [hunk 1]
```diff
@@ -363,15 +363,12 @@ func (stc *ScatterConn) StreamExecuteMulti(
 				qs    queryservice.QueryService
 			)
 			transactionID := info.transactionID
+			reservedID := info.reservedID
```
Comment (systay) [inline go/vt/vtgate/scatter_conn.go:366]: not directly related to your changes, but we need to look over the mutex at the top of the method and check if we are using it correctly. Does not look right to me after skimming the code

File: go/vt/vttablet/endtoend/framework/client.go [hunk 2]
```diff
@@ -238,21 +238,13 @@ func (client *QueryClient) StreamExecute(query string, bindvars map[string]*quer
...
-		options,
-		func(res *sqltypes.Result) error {
-			if result.Fields == nil {
-				result.Fields = res.Fields
-			}
-			result.Rows = append(result.Rows, res.Rows...)
-			return nil
-		},
-	)
+	err := client.server.StreamExecute(client.ctx, client.target, query, bindvars, 0, 0, options, func(res *sqltypes.Result) error {
```
Comment (systay) [inline go/vt/vttablet/endtoend/framework/client.go:241]: nit: these lines become too long after putting them all into a single line. 133 characters is a lot

File: go/vt/vttablet/grpctabletconn/conn.go [hunk 3]
```diff
@@ -832,6 +833,77 @@ func (conn *gRPCQueryClient) ReserveBeginExecute(ctx context.Context, target *qu
 	return sqltypes.Proto3ToResult(reply.Result), reply.TransactionId, reply.ReservedId, conn.tablet.Alias, nil
 }
 
+//ReserveBeginStreamExecute implements the queryservice interface
```
Comment (systay) [inline go/vt/vttablet/grpctabletconn/conn.go:836]: space before comments

File: go/vt/vttablet/grpctabletconn/conn.go [hunk 4]
```diff
@@ -863,6 +935,74 @@ func (conn *gRPCQueryClient) ReserveExecute(ctx context.Context, target *querypb
 	return sqltypes.Proto3ToResult(reply.Result), reply.ReservedId, conn.tablet.Alias, nil
 }
 
+//ReserveStreamExecute implements the queryservice interface
+func (conn *gRPCQueryClient) ReserveStreamExecute(ctx context.Context, target *querypb.Target, preQueries []string, sql string, bindVariables map[string]*querypb.BindVariable, transactionID int64, options *querypb.ExecuteOptions, callback func(*sqltypes.Result) error) (reservedID int64, alias *topodatapb.TabletAlias, err error) {
```
Comment (systay) [inline go/vt/vttablet/grpctabletconn/conn.go:939]: there is a lot of copy and paste with the code block above :point_up: 

any chance we could extract common code so we don't have to duplicate it?

File: go/vt/vttablet/queryservice/queryservice.go [hunk 5]
```diff
@@ -81,7 +81,7 @@ type QueryService interface {
 	// Query execution
 	Execute(ctx context.Context, target *querypb.Target, sql string, bindVariables map[string]*querypb.BindVariable, transactionID, reservedID int64, options *querypb.ExecuteOptions) (*sqltypes.Result, error)
 	// Currently always called with transactionID = 0
```
Comment (systay) [inline go/vt/vttablet/queryservice/queryservice.go:83]: comment is lying

File: go/vt/vttablet/queryservice/wrapped.go [hunk 6]
```diff
@@ -190,10 +190,11 @@ func (ws *wrappedService) Execute(ctx context.Context, target *querypb.Target, q
 	return qr, err
 }
 
-func (ws *wrappedService) StreamExecute(ctx context.Context, target *querypb.Target, query string, bindVars map[string]*querypb.BindVariable, transactionID int64, options *querypb.ExecuteOptions, callback func(*sqltypes.Result) error) error {
-	return ws.wrapper(ctx, target, ws.impl, "StreamExecute", false, func(ctx context.Context, target *querypb.Target, conn QueryService) (bool, error) {
+func (ws *wrappedService) StreamExecute(ctx context.Context, target *querypb.Target, query string, bindVars map[string]*querypb.BindVariable, transactionID int64, reservedID int64, options *querypb.ExecuteOptions, callback func(*sqltypes.Result) error) error {
```
Comment (systay) [inline go/vt/vttablet/queryservice/wrapped.go:193]: public methods should have comments, right?

File: go/vt/vttablet/sandboxconn/sandboxconn.go [hunk 7]
```diff
@@ -400,7 +400,7 @@ func (sbc *SandboxConn) BeginStreamExecute(ctx context.Context, target *querypb.
 	if err != nil {
 		return 0, nil, err
 	}
-	err = sbc.StreamExecute(ctx, target, sql, bindVariables, transactionID, options, callback)
+	err = sbc.StreamExecute(ctx, target, sql, bindVariables, transactionID, 0, options, callback)
```
Comment (systay) [inline go/vt/vttablet/sandboxconn/sandboxconn.go:403]: don't we wan to pass on the reserved id?

File: go/vt/vttablet/tabletserver/tabletserver.go [hunk 8]
```diff
@@ -935,7 +952,7 @@ func (tsv *TabletServer) BeginStreamExecute(
 		return 0, nil, err
 	}
 
-	err = tsv.StreamExecute(ctx, target, sql, bindVariables, transactionID, options, callback)
+	err = tsv.StreamExecute(ctx, target, sql, bindVariables, transactionID, 0, options, callback)
```
Comment (systay) [inline go/vt/vttablet/tabletserver/tabletserver.go:955]: hmm.... should reserved id always be zero here? I'm confused

File: go/vt/vttablet/tabletserver/tabletserver.go [hunk 9]
```diff
@@ -1190,15 +1207,61 @@ func (tsv *TabletServer) ReserveBeginExecute(ctx context.Context, target *queryp
...
+			logStats.ReservedID = connID
+			return nil
+		},
+	)
+
+	if err != nil {
+		return 0, 0, nil, err
+	}
+
+	err = tsv.StreamExecute(ctx, target, sql, bindVariables, connID, 0, options, callback)
```
Comment (systay) [inline go/vt/vttablet/tabletserver/tabletserver.go:1244]: I think we should pass on the connID on both the txId and the reservedId

File: proto/query.proto [hunk 10]
```diff
@@ -674,6 +675,7 @@ message BeginStreamExecuteRequest {
   BoundQuery query = 4;
   ExecuteOptions options = 5;
   repeated string pre_queries = 6;
+  int64 reserved_id = 7;
```
Comment (systay) [inline proto/query.proto:678]: should this not be `transaction_id`?

**Review by systay — COMMENTED** [review]
:tada: 

so nice to unify the two modes

found a few issues that should be addressed, I think

File: go/vt/vtgate/scatter_conn.go (hunk 1 above)
Comment (harshit-gangal) [reply go/vt/vtgate/scatter_conn.go:366]: `ExecuteMultiShard` also doing the same thing, we can look at this together.

File: go/vt/vttablet/grpctabletconn/conn.go (hunk 4 above)
Comment (harshit-gangal) [reply go/vt/vttablet/grpctabletconn/conn.go:939]: the code looks very different to me, not sure how to extract it into common code and would it give any better readability.

File: go/vt/vttablet/queryservice/wrapped.go (hunk 6 above)
Comment (harshit-gangal) [reply go/vt/vttablet/queryservice/wrapped.go:193]: I think if the struct has a comment then linter does not complain at function level comments.

File: go/vt/vttablet/sandboxconn/sandboxconn.go (hunk 7 above)
Comment (harshit-gangal) [reply go/vt/vttablet/sandboxconn/sandboxconn.go:403]: nice catch, BeginStreamExecute needs reserved connection handling.

File: go/vt/vttablet/tabletserver/tabletserver.go (hunk 8 above)
Comment (harshit-gangal) [reply go/vt/vttablet/tabletserver/tabletserver.go:955]: no, `BeginStreamExecute` is not handling reserved connection, need code change.

File: go/vt/vttablet/grpctabletconn/conn.go (hunk 4 above)
Comment (systay) [reply go/vt/vttablet/grpctabletconn/conn.go:940]: Ok, not easy to do without generics. one of these days...

File: go/vt/vtgate/scatter_conn.go (hunk 1 above)
Comment (systay) [reply go/vt/vtgate/scatter_conn.go:366]: nah, `ExecuteMultiShard` actually needs its mutex. I don't think `StreamExecuteMulti` needs it though

**Review by systay — APPROVED** [review]

**Pull request merged**